schedule/reschedule.py

```diff
@@ -298,88 +221,20 @@ def reschedule_background(did, recent=False, filter_flag=False, filtered_cids={}
-    last_date = None
-    last_s = None
-    last_rating = None
-    last_kind = None
-    s = None
-    d = None
-    rating = None
-    revlogs = filter_revlogs(mw.col.card_stats_data(cid).revlog)
-    reps = len(revlogs)
-    for i, revlog in enumerate(reversed(revlogs)):
-        if (
-            i == 0
-            and (revlog.review_kind not in (REVLOG_LRN, REVLOG_RELRN))
-            and not (has_again(revlogs) or has_manual_reset(revlogs))
-        ):
-            break
-        last_s = s
-        rating = revlog.button_chosen
-
-        if (
-            last_kind is not None
-            and last_kind in (REVLOG_REV, REVLOG_RELRN)
-            and revlog.review_kind == REVLOG_LRN
-        ):
-            # forget card
-            last_date = None
-            last_s = None
-            s = None
-            d = None
-        last_kind = revlog.review_kind
-
-        if last_kind == REVLOG_RESCHED:
-            if revlog.ease != 0:
-                # set due date
-                continue
-            else:
-                # forget card
-                last_date = None
-                last_s = None
-                s = None
-                d = None
-                continue
-
-        if last_date is None:
-            again_s = fsrs.init_stability(1)
-            hard_s = fsrs.init_stability(2)
-            good_s = fsrs.init_stability(3)
-            easy_s = fsrs.init_stability(4)
-            d = fsrs.init_difficulty(rating)
-            s = fsrs.init_stability(rating)
-            last_date = datetime.fromtimestamp(revlog.time - rollover * 60 * 60)
-            last_rating = rating
-        else:
-            elapsed_days = (
-                datetime.fromtimestamp(revlog.time - rollover * 60 * 60).toordinal()
-                - last_date.toordinal()
-            )
-            if elapsed_days <= 0:
-                continue
-            r = power_forgetting_curve(elapsed_days, s)
-            fsrs.elapsed_days = elapsed_days
-            again_s = fsrs.next_forget_stability(fsrs.next_difficulty(d, 1), s, r)
-            hard_s = fsrs.next_recall_stability(fsrs.next_difficulty(d, 2), s, r, 2)
-            good_s = fsrs.next_recall_stability(fsrs.next_difficulty(d, 3), s, r, 3)
-            easy_s = fsrs.next_recall_stability(fsrs.next_difficulty(d, 4), s, r, 4)
-            d = fsrs.next_difficulty(d, rating)
-            s = (
-                fsrs.next_recall_stability(d, s, r, rating)
-                if rating > 1
-                else fsrs.next_forget_stability(d, s, r)
-            )
-            last_date = datetime.fromtimestamp(revlog.time - rollover * 60 * 60)
-            last_rating = rating
-
-    if rating is None or s is None:
-        return None
+def reschedule_card(cid, fsrs: FSRS, recompute=False):
+    card = mw.col.get_card(cid)
+    if recompute:
+        memory_state = mw.col.compute_memory_state(cid)
+        s = memory_state.stability
+        d = memory_state.difficulty
```

Comment:
These will be None in the case of new cards, so initial d/s still needs to be dealt with.

Reply:
I have exclude new cards with:

```
        WHERE queue IN ({QUEUE_TYPE_REV}, {QUEUE_TYPE_REV}, {QUEUE_TYPE_DAY_LEARN_RELEARN})
```